### ISSUE TITLE
Disaster recovery fails due to ResourceQuota admission controller rejecting init container in etcd pods

### ISSUE THREAD

**Issue by mlowery** [coreos/etcd-operator#1182]
Preconditions:

* Specify requests and limits (via `spec.pod.resources`) in the etcd cluster object as described [here](https://github.com/coreos/etcd-operator/blob/master/doc/user/spec_examples.md)
* Have a `NotBestEffort`-scoped quota in place (see [here](https://kubernetes.io/docs/concepts/policy/resource-quotas/))

It seems the problem is that the init containers added to the etcd pods during recovery have no requests and limits specified. So the recovery etcd pod is rejected by the `ResourceQuota` admission controller since it wants all containers (regular and init) to have resources specified.

Possible solutions:

* Accept limits and requests on these init containers from the user
* Hard-code them
* Document that a `LimitRange` may be required if such a quota is in place

**Comment by hongchaodeng** [comment]
Can you provide more details on how to reproduce it?

**Comment by mlowery** [comment]
@hongchaodeng:

1. Enable [`ResourceQuota` admission controller](https://kubernetes.io/docs/admin/admission-controllers/#resourcequota).
1. Create a `NotBestEffort`-scoped quota in the namespace:

    ```yaml
    apiVersion: v1
    kind: ResourceQuota
    metadata:
      name: high-qos-limit-requests
    spec:
      hard:
        limits.cpu: "8"
        limits.memory: 24Gi
        pods: "8"
        requests.cpu: "8"
        requests.memory: 24Gi
      scopes:
      - NotBestEffort
    ```
2. Specify `spec.pod.resources` in etcd cluster object and let the cluster stabilize.
3. Trigger a disaster recovery. I stumbled upon a disaster recovery accidentally but I assume it can be triggered by deleting all of the etcd pods.
4. Tailing etcd-operator logs shows:

    ```
    msg="all etcd pods are dead. Trying to recover from a previous backup"
    msg="fail to do disaster recovery: failed to create seed member (mlowery-etcd-0002): pods \"mlowery-etcd-0002\" is forbidden: failed quota: high-qos-limit-requests:
    ```

**Comment by mlowery** [comment]
@hongchaodeng:

A `LimitRange` such as the following fixed my problem:

```yaml
apiVersion: v1
kind: LimitRange
metadata:
  name: limits
spec:
  limits:
  - default:
      cpu: 100m
      memory: 128Mi
    defaultRequest:
      cpu: 100m
      memory: 128Mi
    type: Container
```

However, this was only done because it was easier than modifying the etcd-operator code.

**Comment by xiang90** [comment]
@mlowery probably we can set a small default requested resource for the initial container. it should require a lot of resource anyway. 256MB and 200m seems fine.

**Comment by nabadger** [comment]
I believe I hit this issue when setting up the initial cluster as well (maybe it shares the same code as restore?).

Some logs of the etcd-operator:

> time="2018-01-09T15:56:54Z" level=info msg="etcd-operator Version: 0.8.0" 
time="2018-01-09T15:56:54Z" level=info msg="Git SHA: 14aa69c" 
time="2018-01-09T15:56:54Z" level=info msg="Go Version: go1.9.2" 
time="2018-01-09T15:56:54Z" level=info msg="Go OS/Arch: linux/amd64" 
time="2018-01-09T15:57:11Z" level=info msg="Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"examples", Name:"etcd-operator", UID:"0c3c2465-f555-11e7-9ecc-525400a46477", APIVersion:"v1", ResourceVersion:"3474146", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' etcd-operator-7d46996f58-bwclv became leader" 
time="2018-01-09T15:57:45Z" level=info msg="creating cluster with Spec:" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="{" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="    "size": 3," cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="    "repository": "quay.io/coreos/etcd"," cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="    "version": "3.2.13"," cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="    "pod": {" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="        "resources": {" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="            "limits": {" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="                "cpu": "100m"," cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="                "memory": "128Mi"" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="            }," cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="            "requests": {" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="                "cpu": "30m"," cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="                "memory": "50Mi"" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="            }" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="        }" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="    }" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=info msg="}" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=error msg="cluster failed to setup: failed to create seed member (etcd-cluster-0000): pods "etcd-cluster-0000" is forbidden: failed quota: default: must specify limits.cpu,limits.memory,requests.cpu,requests.memory" cluster-name=etcd-cluster pkg=cluster 
time="2018-01-09T15:57:45Z" level=warning msg="fail to handle event: ignore failed cluster (etcd-cluster). Please delete its CR" pkg=controller

**Comment by hongchaodeng** [comment]
Will fix this

**Comment by hongchaodeng** [comment]
We will add `spec.pod.resources` into the init container, which we should anyway.

Also check https://kubernetes.io/docs/concepts/policy/resource-quotas/:

> If quota is enabled in a namespace for compute resources like cpu and memory, users must specify requests or limits for those values; otherwise, the quota system may reject pod creation. Hint: Use the LimitRange admission controller to force defaults for pods that make no compute resource requirements. See the walkthrough for an example to avoid this problem.

It's still recommended to use LimitRange to set defaults because there are other resources like "ephemeral-storage" and new types added in the future.

**Comment by hongchaodeng** [comment]
fixed in #1845
